### DISCUSSION TITLE
ADFS as external IdP

### DISCUSSION THREAD

**Discussion by megge-dream** [zitadel/zitadel#5937] [General]
Hi!
I'm trying to use ADFS as external IdP for Zitadel. I created Generic OIDC provider for this, but when I try to log in with ADFS credentials into test web application ([https://github.com/zitadel/react-user-authentication](https://github.com/zitadel/react-user-authentication)), I'm getting 401 error.

I did some research and find out that ADFS returns 401 error in response to `/userinfo` endpoint - `error="invalid_token", error_description="MSIS9921: Received invalid UserInfo request. Audience 'microsoft:identityserver:XXXX' in the access token is not same as the identifier of the UserInfo relying party trust 'urn:microsoft:userinfo'`. As a workaround to it I tried to create Generic OAuth Provider instead and add param `resource=urn%3Amicrosoft%3Auserinfo` to Authorization Endpoint (as described [here](https://serverfault.com/questions/1109792/setting-up-oidc-with-adfs-invalid-userinfo-request)). Now I get 200 response, but don't have any info about user profile and email.
As I understand ADFS `/userinfo` endpoint only returns the subject claim, and all additional claims can be provided in ID token ([from the ADFS FAQ](https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/overview/ad-fs-faq#i-m-trying-to-get-more-claims-on-the-userinfo-endpoint--but-it-s-only-returning-subject--how-can-i-get-more-claims-))

So is there a way to use ADFS as external IdP? Maybe I can set some settings to make Zitadel get all info from id token and/or don't go to `/userinfo` endpoint at all?

**Comment by fforootd** [comment]
Just to understand that I understand this correctly.

You attached a React Web App to ZITADEL and configured a Generic OIDC IdP connected to an ADFS with OIDC,  this did not work with the error you saw.

When you switched to Generic OAuth in ZITADEL you got the 200 but without the claims, right?

What happens if you use the generic OAuth and add the scopes `openid profile email`?

**Reply by fforootd** [reply]
Also CC @livio-a maybe you have some thoughts on this as well.

**Reply by megge-dream** [reply]
Yes, you described everything correct.
I did use `openid profile email` scopes for both cases (generic OAuth and generic OIDC), and I set up custom claims for this scopes in ADFS as well.
I get `profile email` info in accessToken from Authorization Endpoint when I **don't** use param 
 `resource=urn%3Amicrosoft%3Auserinfo` in request to it, but `/userinfo` return 401 for this token.
If I use generic OAuth with `resource=urn%3Amicrosoft%3Auserinfo` in Authorization Endpoint, I don't get info in accessToken, and `/userinfo` return 200 for this token with only subject field.
 I think if it would be possible not to go to `/userinfo` endpoint at all, I probably could parse accessToken in Post Authentication Action and get all needed information from it.
